include/revng/Model/NameBuilder.h

```diff
@@ -199,6 +202,189 @@ class NameBuilder {
+
+      bool IsLocationUnique = EmittedLocations.emplace(std::move(AsStr)).second;
+      if (not IsLocationUnique) {
+        // This location was used for a different variable already,
+        // emit an *untagged* automatic name.
+        return automaticName(AutomaticPrefix, false);
+      }
+
+      auto Comparator = [&](const model::LocalIdentifier &Identifier) {
+        return Identifier.Location() == UserLocationSet;
+      };
+      auto Iterator = std::ranges::find_if(KnownNames, Comparator);
+      if (Iterator == KnownNames.end()) {
+        // There's nothing in the model for the current location,
+        // emit a *tagged* automatic name.
+        return automaticName(AutomaticPrefix, true);
+      }
+
+      bool IsNameUnique = EmittedNames.emplace(Iterator->Name()).second;
+      if (not IsNameUnique) {
+        // This name was already emitted, fall back on the automatic name.
+        auto Result = automaticName(AutomaticPrefix, true);
+        Result.Warning = "Name `" + Iterator->Name()
+                         + "` cannot be used more than once, so it was "
+                           "replaced by an automatic one (`"
+                         + Result.Name + "`).";
+        return Result;
+      }
+
+      revng_assert(Parent);
+      if (llvm::Error Reason = Parent->isNameReserved(Iterator->Name())) {
+        // Current name cannot be used as its reserved for something else.
+        auto Result = automaticName(AutomaticPrefix, true);
+        Result.Warning = "Name `" + Iterator->Name()
+                         + "` is not valid, so it was replaced by an automatic "
+                           "one (`"
+                         + Result.Name + "`) because "
+                         + revng::unwrapError(std::move(Reason)) + ".";
+        return Result;
+
+      } else {
+        // We don't want existing automatic names to "shift" when a rename
+        // happens, as such we have to advance the index even if no automatic
+        // name is produced for it.
+        uint64_t CurrentIndex = NextIndex++;
+
+        return { Iterator->Name(), CurrentIndex, true };
+      }
+    }
+
+    [[nodiscard]] std::set<llvm::StringRef>
+    homelessNamesImpl(RangeOf<const model::LocalIdentifier &> auto R) const {
+      return R | std::views::filter([this](const auto &V) {
+               return not EmittedNames.contains(V.Name());
+             })
+             | std::views::transform([](const auto &V) { return V.Name(); })
+             | revng::to<std::set<llvm::StringRef>>();
+    }
+  };
+
+  struct VariableNameBuilder : public CountingNameBuilder {
+    const model::Function *Function = nullptr;
+
+    VariableNameBuilder() = default;
+    VariableNameBuilder(const NameBuilder &Parent,
+                        const model::Function &Function) :
+      CountingNameBuilder(Parent), Function(&Function) {}
+
+    auto name(RangeOf<MetaAddress> auto const &UserLocationSet) {
```

Comment:
The function prototype act as documentation. Do not abuse `auto` as return type.

Reply:
Done.